src/Microsoft.OpenApi/Services/OpenApiComparerBase.cs

```diff
@@ -138,6 +166,125 @@ internal void Compare<TEnum>(Enum source, Enum target, ComparisonContext compari
             }
         }
 
+        /// <summary>
+        /// Compares <see cref="OpenApiReference"/> object.
+        /// </summary>
+        /// <param name="sourceReference">The source.</param>
+        /// <param name="targetReference">The target.</param>
+        /// <param name="comparisonContext">The context under which to compare the objects.</param>
+        internal void Compare<TReference>(
+            OpenApiReference sourceReference,
+            OpenApiReference targetReference,
+            ComparisonContext comparisonContext)
+        {
+            if (sourceReference == null && targetReference == null)
+            {
+                return;
+            }
+
+            if (sourceReference == null || targetReference == null)
+            {
+                comparisonContext.AddOpenApiDifference(
+                    new OpenApiDifference
+                    {
+                        OpenApiDifferenceOperation = OpenApiDifferenceOperation.Update,
+                        SourceValue = sourceReference,
+                        TargetValue = targetReference,
+                        OpenApiComparedElementType = typeof(OpenApiReference),
+                        Pointer = comparisonContext.PathString
+                    });
+
+                return;
+            }
+
+            if (sourceReference.Id != targetReference.Id || sourceReference.Type != targetReference.Type)
+            {
+                WalkAndAddOpenApiDifference(
+                    comparisonContext,
+                    OpenApiConstants.DollarRef,
+                    new OpenApiDifference
+                    {
+                        OpenApiDifferenceOperation = OpenApiDifferenceOperation.Update,
+                        SourceValue = sourceReference,
+                        TargetValue = targetReference,
+                        OpenApiComparedElementType = typeof(OpenApiReference)
+                    });
+
+                return;
+            }
+
+            var source = (TReference) comparisonContext.SourceDocument.ResolveReference(
+                sourceReference);
+
+            var target = (TReference) comparisonContext.TargetDocument.ResolveReference(
+                targetReference);
+
+            comparisonContext
+                .GetComparer<TReference>()
+                .Compare(source, target, comparisonContext);
+        }
```

Comment:
This seems like something that should be in its own class OpenApiReferenceComparer rather than the base class

Reply:
I tried that but didn't go it b/c even if i add as separate comparer class like below

public class OpenApiReferenceComparer<T> : OpenApiComparerBase<OpenApiReference> where T : IOpenApiReferenceable

It can't be registered under OpenApiComparerFactory, as there the type of reference is an enum and can't register varies reference comparer based on the value of enum,

That said i will still add the comparer class and reference it directly in the comparers where we diff reference.


---
In reply to: [223511051](https://github.com/Microsoft/OpenAPI.NET/pull/328#discussion_r223511051) [](ancestors = 223511051)